### PR TITLE
Rework _DeepMatcher so that it can display details of match failures.

### PR DESCRIPTION
**Pull request by stereotype441**
Previously, if the user tried to do a deep match where the expectation
included a complex matcher (such as a "having" matcher), the failure
message would just say "failed to match ..."; it wouldn't call on the
expectation's matcher to explain why the match failed.

`_DeepMatcher` now stores a `_Mismatch` object in the matchState
map (under the key `mismatch`) recording in detail the location
of the mismatch, the actual and expected values, and if
necessary, a callback that describes the problem in more detail.
This replaces the old `reason` object, which was a String.  Using
a `_Mismatch` object allows us to pass the `verbose` flag through
when calling on the inner expectation's matcher to explain why
the match failed.  It also avoids unnecessary string operations
if no description is needed.

### CODE REVIEW

**Comment by natebosch** [comment]
+1 on a changelog entry - the current commit message would easily translate to a good changelog entry. I think we should have an expanded commit message that has a brief description of the design change - you could edit the first comment in this thread and we could copy it into the commit message when we squash and merge.

Overall I think the change looks good, I'll take another more detailed pass through the code.

cc @davidmorgan - this breaks a test in `built_value_test` that hardcodes a description coming from this matcher. It should be easy to adjust. You won't see the change though until we publish here as well as `package:test`.

File: lib/src/equals_matcher.dart [hunk 1]
```diff
@@ -214,44 +243,24 @@ class _DeepMatcher extends Matcher {
...
-    return reason;
   }
 
   @override
-  bool matches(item, Map matchState) =>
-      _match(_expected, item, matchState) == null;
+  bool matches(Object actual, Map matchState) {
+    var mismatch = _recursiveMatch(_expected, actual, '', 0);
+    if (mismatch == null) return true;
+    addStateInfo(matchState, {'mismatch': mismatch});
```
Comment (natebosch) [inline lib/src/equals_matcher.dart:261]: This is a detail that would be good to describe in the commit message.

File: lib/src/equals_matcher.dart [hunk 2]
```diff
@@ -260,16 +269,48 @@ class _DeepMatcher extends Matcher {
...
     return mismatchDescription;
   }
 }
+
+class _Mismatch {
+  final String location;
+
+  final Object actual;
+
+  final void Function(Description description, bool verbose) describeProblem;
```
Comment (natebosch) [inline lib/src/equals_matcher.dart:306]: [nit] We don't have to specify a variable name in a function type, and where it matches the type I prefer to omit it.

```suggestion
  final void Function(Description, bool verbose) describeProblem;
```

File: lib/src/equals_matcher.dart [hunk 3]
```diff
@@ -260,16 +269,48 @@ class _DeepMatcher extends Matcher {
...
+
+  final bool instead;
```
Comment (natebosch) [inline lib/src/equals_matcher.dart:308]: Can you add a doc comment here?

File: lib/src/equals_matcher.dart [hunk 4]
```diff
@@ -260,16 +269,48 @@ class _DeepMatcher extends Matcher {
...
+      } else {
+        mismatch.describeProblem(mismatchDescription, verbose);
+      }
     }
     return mismatchDescription;
   }
 }
+
+class _Mismatch {
+  final String location;
```
Comment (natebosch) [inline lib/src/equals_matcher.dart:302]: Can you add a doc comment? I assume this is something like "A human readable description of the location within a collection."?

**Comment by davidmorgan** [comment]
@natebosch Noted, thanks!

**Comment by natebosch** [comment]
When I look at the error in `built_value_test` I do notice that there appears to be a bug somewhere here:

```
Which: at location [\'name\'] is \'foo\' instead of  \'bar\'\n'
```

Notice 2 spaces following `instead of`. That looks like it should be 1. We will need to find where the extra is coming from.

File: lib/src/equals_matcher.dart [hunk 5]
```diff
@@ -260,16 +269,48 @@ class _DeepMatcher extends Matcher {
...
+    var mismatch = matchState['mismatch'] as _Mismatch;
+    if (mismatch.location.isNotEmpty) {
+      mismatchDescription
+          .add('at location ')
+          .add(mismatch.location)
+          .add(' is ')
+          .addDescriptionOf(mismatch.actual);
+      if (mismatch.describeProblem != null) {
+        mismatchDescription
+            .add(' ${mismatch.instead ? 'instead of ' : 'which'} ');
```
Comment (natebosch) [inline lib/src/equals_matcher.dart:281]: This looks like where the extra space is coming from.

```suggestion
            .add(' ${mismatch.instead ? 'instead of' : 'which'} ');
```

It could be that this string isn't currently covered by any tests here?

**Comment by natebosch** [comment]
@stereotype441 - is this still something you want to push on?

**Comment by stereotype441** [comment]
> Worth a changelog entry!

Done!

**Comment by stereotype441** [comment]
> +1 on a changelog entry - the current commit message would easily translate to a good changelog entry. I think we should have an expanded commit message that has a brief description of the design change - you could edit the first comment in this thread and we could copy it into the commit message when we squash and merge.

Ok, I've edited the commit message.  Let me know if there's more information you'd like there.

File: lib/src/equals_matcher.dart (hunk 1 above)
Comment (stereotype441) [reply lib/src/equals_matcher.dart:261]: Done

File: lib/src/equals_matcher.dart (hunk 2 above)
Comment (stereotype441) [reply lib/src/equals_matcher.dart]: Done

File: lib/src/equals_matcher.dart (hunk 3 above)
Comment (stereotype441) [reply lib/src/equals_matcher.dart:318]: Done

File: lib/src/equals_matcher.dart (hunk 4 above)
Comment (stereotype441) [reply lib/src/equals_matcher.dart:304]: Done

File: lib/src/equals_matcher.dart (hunk 5 above)
Comment (stereotype441) [reply lib/src/equals_matcher.dart]: Fixed.  It turns out that this string *is* covered by tests, it’s just that the error messages are all matched ignoring whitespace (see `shouldFail` in `test/test_utils.dart`).  It seems a little silly to add a test just to make sure this whitespace issue doesn’t regress, but I can do so if you want.

**Comment by stereotype441** [comment]
> @stereotype441 - is this still something you want to push on?

Yes, sorry for dropping the ball on this.  PTAL and let me know what you think of the revised CL.

Note that the Travis failures are due to the fact that the dev SDK has a slightly different phrasing for range error messages (see https://github.com/dart-lang/sdk/commit/10e8d5d1a698fac27b6be60b0ae13c519a19f68a).

**Comment by natebosch** [comment]
I think we should update the test to use the dev SDK phrasing and skip the test on older SDKs.

**Comment by stereotype441** [comment]
> Overall this LGTM once the tests are fixed.
> 
> We need to figure out timing on when to release this. @jakemac53 - do we need the NNBD version of matcher published this week?

Ok, I fixed the tests.  Travis is now failing due to issues with `unawaited` (which I believe @jakemac53 and @bwilkerson are working on).  I'm happy to wait until those issues are sorted out before landing this.  (Which works well anyway since @natebosch is on vacation until Monday)

**Comment by kevmoo** [comment]
@stereotype441 – we could bump the min SDK to 2.7 for this package.

Reasonable, @natebosch ?

**Comment by stereotype441** [comment]
> @stereotype441 – we could bump the min SDK to 2.7 for this package.
> 
> Reasonable, @natebosch ?

What would be the benefit of bumping the min SDK to 2.7?  If you're trying to solve the issue with the tests, that won't work because the error text changed in 2.9.0-13.0-dev.  (Besides, bumping the min SDK just to solve an issue with an overly specific test seems unnecessary; better to broaden the test so that it passes on older SDKs rather than lock people with older SDKs out of upgrading this package)

**Comment by kevmoo** [comment]
Folks generally DON'T run old SDKs. And we don't generally support them
beyond one release.

It's okay to make our lives easier IMHO

**Comment by stereotype441** [comment]
> Folks generally DON'T run old SDKs. And we don't generally support them beyond one release. It's okay to make our lives easier IMHO

Agreed that it's ok to make our lives easier, but (a) the problem with the tests is already fixed, and (b) even if it hadn't been fixed, bumping the min SDK to 2.7 would not have fixed it.

If you want to make a separate PR to bump the min SDK to 2.7 I have no objection, but there's no relation to this PR.
